src/build/calibrate/calibrate_national.jl

```diff
@@ -0,0 +1,183 @@
+function calibrate_national(
+    dataset::String,
+    io::Dict,
+    set::Dict;
+    save_build::Bool=SLiDE.DEFAULT_SAVE_BUILD,
+    overwrite::Bool=SLiDE.DEFAULT_OVERWRITE,
+    penalty_nokey::AbstractFloat=SLiDE.DEFAULT_PENALTY_NOKEY,
+)
+    subset = "calibrate"
+
+    # If there is already calibration data, read it and return.
+    d_read = SLiDE.read_build(dataset, subset; overwrite=overwrite)
+    !(isempty(d_read)) && (return d_read)
+    
+    # Initialize a DataFrame to contain results and do the calibration iteratively.
+    cal = Dict(k => DataFrame() for k in list_parameters!(set, :calibrate))
+
+    for year in set[:yr]
+        cal_yr = calibrate_national(io, set, year; penalty_nokey=penalty_nokey)
+        [cal[k] = [cal[k]; cal_yr[k]] for k in keys(cal_yr)]
+    end
+    
+    # If no DataFrame was returned by the annual calibrations, replace this with the input.
+    [cal[k] = io[k] for (k,df) in cal if isempty(df)]
+
+    SLiDE.write_build!(dataset, subset, cal; save_build=save_build)
+    return cal
+end
+
+
+function calibrate_national(
+    io::Dict,
+    set::Dict,
+    year::Int;
+    penalty_nokey::AbstractFloat=SLiDE.DEFAULT_PENALTY_NOKEY,
+    # multipliers for lower and upper bound relative
+    # to each respective variables reference parameter
+    lower_bound = SLiDE.DEFAULT_CALIBRATE_LOWER_BOUND,
+    upper_bound = SLiDE.DEFAULT_CALIBRATE_UPPER_BOUND,
+)
+    @info("Calibrating $year data")
+
+    # Save relevant sets to simplify things a little.
+    S, G, M, VA, FD = set[:s], set[:g], set[:m], set[:va], set[:fd]
+    
+    SLiDE._calibration_set!(set; final_demand=true, value_added=true)
+    d = SLiDE._national_calibration_input(io, set, year;
+        lower_bound=lower_bound,
+        upper_bound=upper_bound,
+    )
```

Comment:
pre-calculates upper and lower bounds and adds them to `d` (detonated `:k_lb` and `:k_ub`. for a key `:k`)